/*
 * Copyright (C) 2025 ROS-Industrial Consortium Asia Pacific
 * Advanced Remanufacturing and Technology Centre
 * A*STAR Research Entities (Co. Registration No. 199702110H)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#ifndef VDA5050_MSGS__JSON_UTILS__NODE_POSITION_HPP_
#define VDA5050_MSGS__JSON_UTILS__NODE_POSITION_HPP_

#include <string>

#include <nlohmann/json.hpp>

#include "vda5050_msgs/msg/node_position.hpp"

namespace vda5050_msgs {

namespace msg {

//=============================================================================
/// \brief Convert a vda5050_msgs::msg::NodePosition object to a
/// nlohmann::json object
///
/// \param j Reference to the JSON object to be populated
/// \param msg Reference to the message object to serialize
inline void to_json(nlohmann::json& j, const NodePosition& msg)
{
  j["x"] = msg.x;
  j["y"] = msg.y;
  j["mapId"] = msg.map_id;

  if (!msg.theta.empty())
  {
    j["theta"] = msg.theta.front();
  }

  if (!msg.allowed_deviation_x_y.empty())
  {
    j["allowedDeviationXY"] = msg.allowed_deviation_x_y.front();
  }

  if (!msg.allowed_deviation_theta.empty())
  {
    j["allowedDeviationTheta"] = msg.allowed_deviation_theta.front();
  }

  if (!msg.map_description.empty())
  {
    j["mapDescription"] = msg.map_description.front();
  }
}

//=============================================================================
/// \brief Populate a vda5050_msgs::msg::NodePosition object from a
/// nlohmann::json object
///
/// \param j Reference to the JSON object containing serialized data
/// \param msg Reference to the message object to populate
inline void from_json(const nlohmann::json& j, NodePosition& msg)
{
  msg.x = j.at("x").get<double>();
  msg.y = j.at("y").get<double>();
  msg.map_id = j.at("mapId").get<std::string>();

  if (j.contains("theta"))
  {
    msg.theta.push_back(j.at("theta").get<double>());
  }

  if (j.contains("allowedDeviationXY"))
  {
    msg.allowed_deviation_x_y.push_back(
      j.at("allowedDeviationXY").get<double>());
  }

  if (j.contains("allowedDeviationTheta"))
  {
    msg.allowed_deviation_theta.push_back(
      j.at("allowedDeviationTheta").get<double>());
  }

  if (j.contains("mapDescription"))
  {
    msg.map_description.push_back(j.at("mapDescription").get<std::string>());
  }
}

}  // namespace msg
}  // namespace vda5050_msgs

#endif  // VDA5050_MSGS__JSON_UTILS__NODE_POSITION_HPP_
